src/main/java/com/bugsnag/BugsnagAppender.java

```diff
@@ -0,0 +1,681 @@
+                    .startsWith("com.bugsnag.delivery.OutputStreamDelivery")
+                    || element.getClassName()
+                    .startsWith("com.bugsnag.delivery.SyncHttpDelivery")) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Gets the throwable from the log event (if any)
+     * @param event The log event
+     * @return The throwable (or null)
+     */
+    private Throwable extractThrowable(ILoggingEvent event) {
+        IThrowableProxy throwableProxy = event.getThrowableProxy();
+        if (throwableProxy instanceof ThrowableProxy) {
+            return ((ThrowableProxy) event.getThrowableProxy()).getThrowable();
+        }
+
+        return null;
+    }
+
+    /**
+     * @return Create a Bugsnag instance with parameters from the logback configuration
+     */
+    private Bugsnag createBugsnag() {
+        Bugsnag bugsnag = Bugsnag.init(apiKey, sendUncaughtExceptions);
+
+        bugsnag.setAutoCaptureSessions(autoCaptureSessions);
+
+        if (appType != null) {
+            bugsnag.setAppType(appType);
+        }
+
+        if (appVersion != null) {
+            bugsnag.setAppVersion(appVersion);
+        }
+
+        if (notifyEndpoint != null) {
+            bugsnag.setEndpoints(notifyEndpoint, sessionEndpoint);
+        }
+
+        if (proxy != null) {
+            bugsnag.setProxy(
+                    new Proxy(
+                            proxy.getType(),
+                            new InetSocketAddress(proxy.getHostname(), proxy.getPort())));
+        }
+
+        if (releaseStage != null) {
+            bugsnag.setReleaseStage(releaseStage);
+        }
+
+        if (timeout > 0) {
+            bugsnag.setTimeout(timeout);
+        }
+
+        if (filteredProperties.size() > 0) {
+            bugsnag.setFilters(filteredProperties.toArray(new String[0]));
+        }
+
+        bugsnag.setIgnoreClasses(ignoredClasses.toArray(new String[0]));
+
+        if (notifyReleaseStages.size() > 0) {
+            bugsnag.setNotifyReleaseStages(notifyReleaseStages.toArray(new String[0]));
+        }
+        bugsnag.setProjectPackages(projectPackages.toArray(new String[0]));
+        bugsnag.setSendThreads(sendThreads);
+
+        return bugsnag;
+    }
+
+    /**
+     * Adds the given key / value to the current thread logging context
+     *
+     * @param key the key to add
+     * @param value the value to add
+     */
+    public static void addThreadMetaData(String tab, String key, Object value) {
+        MDC.put(LOGGING_CONTEXT_THREAD_PREFIX + tab
+                + LOGGING_CONTEXT_TAB_SEPARATOR + key, getStringValue(value));
+    }
+
+    /**
+     * Adds the given key / value to the current thread logging context
+     * to be used for the next report
+     *
+     * @param key the key to add
+     * @param value the value to add
+     */
+    public static void addReportMetaData(String tab, String key, Object value) {
```

Comment:
We should clarify that the metadata will only be added to the next report, then it gets removed.

Reply:
added note